site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,576 @@
+
+The resulting system trace will look like this (using [Jaeger](https://www.jaegertracing.io/) web-ui visualization):
+![jaeger view](/blog/2020-06-10-opentracing/result.png) 
+
+## Naive approach
+
+To add tracing to any API call, we can do the following in our `ServeHTTP` function:
+
+```go
+func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	c := &Context{}
+	// Start root span
+	span, ctx := tracing.StartRootSpanByContext(context.Background(), "apiHandler")
+	// Populate different span fields based on request headers
+	carrier := opentracing.HTTPHeadersCarrier(r.Header)
+	_ = opentracing.GlobalTracer().Inject(span.Context(), opentracing.HTTPHeaders, carrier)
+	ext.HTTPMethod.Set(span, r.Method)
+	ext.HTTPUrl.Set(span, c.App.Path())
+	ext.PeerAddress.Set(span, c.App.IpAddress())
+	span.SetTag("request_id", c.App.RequestId())
+	span.SetTag("user_agent", c.App.UserAgent())
+	// On handler exit: a) in case of an error, add it to the trace b) finish the span
+	defer func() {
+		if c.Err != nil {
+			span.LogFields(spanlog.Error(c.Err))
+			ext.HTTPStatusCode.Set(span, uint16(c.Err.StatusCode))
+			ext.Error.Set(span, true)
+		}
+		span.Finish()
+	}()
+	// Set current context to the one we got from root span, it will be passed down to actual api handlers
+	c.App.SetContext(ctx)
+	// ...
+	// Execute the actual API handler
+	h.HandleFunc(c, w, r)
+}
+```
+
+Next, we'll modify the actual API handler to nest it inside the parent span (we'll use `SearchUsers` as example)
+
+```go
+func (a *App) SearchUsers(props *model.UserSearch, options *model.UserSearchOptions) ([]*model.User, *model.AppError) {
+	// save previous context
+	origCtx := a.ctx
+	// generate new span, nested inside the parent span
+	span, newCtx := tracing.StartSpanWithParentByContext(a.ctx, "app.SearchUsers")
+	// set new context
+	a.ctx = newCtx
+	
+	// log some parameters
+	span.SetTag("searchProps", props)
+
+	// on function exit, restore context and finish the span
+	defer func() {
+		a.ctx = origCtx
+		span.Finish()
+	}()
+
+	// ...
+	// perform actual work
+	// ...
+
+	// in case of an error, add it to the span
+	if err != nil {
+		span.LogFields(spanlog.Error(err))
+		ext.Error.Set(span, true)
+	}
+
+	// return results
+}
+```
+
+Rather straight forward, right? We marked our "entry-point" by creating a root span, populating it with useful context information, then passing the context down the stack and creating new span underneath it.
+
+We could stop right here, because this is all you need to have a working trace! **But**: for a large application like `mattermost-server` wrapping all of the 900+ API handlers in tracing code would be incredible labor intensive AND will create a lot of noise in the source code.
```

Comment:
```suggestion
We could stop right here, because this is all you need to have a working trace! **But** for a large application like `mattermost-server` wrapping all of the 900+ API handlers in tracing code would be incredibly labor intensive and will create a lot of noise in the source code.
```